scripts/port_clash_fixup.py

```diff
@@ -0,0 +1,84 @@
+#!/usr/bin/env python3
+# Copyright 2020 ETH Zurich
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+:mod: scripts.port_clash_fixup
+=========================
+
+Prints existing port clashes and fixes them.
+For every clash, the interface that was modified the first is kept, but
+for all the other ones a new port is found, and the associated
+AP's configuration version is bumped.
+This is intended to be used only once.
+
+Run with python manage.py runscript port_clash_fixup
+"""
+
+
+from collections import defaultdict
+from scionlab.models.core import Host, Interface, LazyPortMap
+from scionlab.defines import DEFAULT_PUBLIC_PORT
+
+
+def run():
+    # get the clashes:
+    clashes = dict()
+    for h in Host.objects.all():
+        c = get_clashes(h)
+        if len(c):
+            clashes[h] = c
+    # print them
+    for k, v in clashes.items():
+        print('host {host} has {clashes} clashes:'.format(
+            host=k, clashes=len(v)))
+        for c in v:
+            print('\tsocket = {socket} IDs = {ids}'.format(socket=c[0], ids=c[1]))
+    # fix them
+    for clashes in clashes.values():
```

Comment:
Just in case something goes wrong unexpectedly, it might make sense to wrap this update part into a `django.db.transaction.atomic` (https://docs.djangoproject.com/en/3.0/topics/db/transactions/#django.db.transaction.atomic).

Reply:
I saw that the vpn script didn't use them and understood we didn't care for transactions. Adding them now.

Reply:
I did add the transactions to the VPN script in the last second. The point is not harden against concurrency but to ensure a known state, i.e. make sure that the script either runs to completion or does not change anything.
So we want the transaction over the _entire_ change block, not each individual `fix_change` call.

Reply:
No actually that makes sense here as each `fix_clash` resolves one clash and when one call would fail, just running the script again would only work on the remaining conflicts :+1: